packages/template-retail-react-app/app/utils/responsive-image.js

```diff
@@ -64,29 +88,90 @@ const mapWidthsToSizes = (widths) => {
         .join(', ')
 }
 
+function padArray(arr) {
+    const l1 = arr.length
+    const l2 = breakpointLabels.length
+    if (l1 < l2) {
+        const lastEntry = arr.at(-1)
+        const amountToPad = l2 - l1
+        return [...arr, ...Array(amountToPad).fill(lastEntry)]
+    }
+    return arr
+}
+
 /**
- * @param {(number[]|string[]|Object)} widths
- * @return {string}
+ * @param {(number[]|[number, number][])} densities
+ * @param {number[][][]} [collector]
+ * @returns {number[][][]}
  */
-const mapWidthsToSrcSet = (widths, dynamicSrc) => {
-    let _widths = isObject(widths) ? widthsAsArray(widths) : widths.slice(0)
-
-    if (_widths.length < breakpointLabels.length) {
-        const lastWidth = _widths[_widths.length - 1]
-        const amountToPad = breakpointLabels.length - _widths.length
+function mapDensities(densities, collector) {
+    const flat = Array.isArray(collector)
+    return densities.reduce(
+        (acc, entry) => {
+            if (!flat && Array.isArray(entry) && Array.isArray(entry.at(0))) {
+                acc.push(mapDensities(entry, []))
+            } else if (
+                Array.isArray(entry) &&
+                typeof entry.at(0) === 'number' &&
+                !Number.isNaN(entry.at(0))
+            ) {
+                const w = entry.at(0)
+                const f = entry.at(1)
+                const result = [w, typeof f === 'number' && !Number.isNaN(f) ? f : w]
+                acc.push(flat ? result : [result])
+            } else if (typeof entry === 'number' && !Number.isNaN(entry)) {
+                const result = [entry, entry]
+                acc.push(flat ? result : [result])
+            } else if (flat) {
+                acc.push(...defaultDensityFactors)
+            } else {
+                acc.push(defaultDensityFactors)
+            }
+            return acc
+        },
+        flat ? collector : []
+    )
+}
 
-        _widths = [..._widths, ...Array(amountToPad).fill(lastWidth)]
+function obtainDensities(densities) {
+    const cache = densitiesCache.get(themeBreakpoints)
+    if (cache?.has(densities)) {
+        return densitiesCache.get(themeBreakpoints).get(densities)
```

Comment:
nit: A bit more readable:

```suggestion
        return cache.get(densities)
```